Allow docks to be closed and opened

kitbdev opened the pull request:
- requires https://github.com/godotengine/godot/pull/88003
- closes https://github.com/godotengine/godot-proposals/issues/8206
- related https://github.com/godotengine/godot-proposals/issues/1441
- related https://github.com/godotengine/godot-proposals/issues/7233

Docks can be closed through their context menu, and reopened through the new `Window` menu.
If a dock is already open, clicking on it in the `Window` menu will focus on it.
Dock layouts can be saved with closed tabs. This helps with some of the use cases in https://github.com/godotengine/godot-proposals/issues/1441.

Added dock slot `DOCK_SLOT_NONE`. Using this with `add_control_to_dock` allows the dock to not be opened immediately, it will be closed by default. When it is opened from the `Window` menu, it will appear floating.

Internally, renamed `add_control_to_dock` to `add_dock` and `remove_control_from_dock` to `remove_dock`. This is because I want to unify the language used better, the Control that is being added and removed *is* the dock, and it is added to a dock slot/dock container. However, I didn't change it in the editor plugin since it would break compat. I want to add it as the new names in EditorInterface in a future PR.

Closed docks aren't unloaded, they are just removed from any TabContainers and hidden.

After https://github.com/godotengine/godot/pull/88081, shortcuts can be added to open docks.

![image](https://github.com/godotengine/godot/assets/10054226/44b3215e-9bcd-4563-954e-d24b8415a8b7)

![image](https://github.com/godotengine/godot/assets/10054226/cdf640c0-d435-47cc-91db-372ad495ed36)

Updated (import closed, history disabled):

![image](https://github.com/godotengine/godot/assets/10054226/fdb58c67-d8e3-4f8e-b30b-01a9892039c2)

kitbdev committed:
Allow docks to be closed

Mickeon commented:
What I can say with confidence is that: 
- The menu could be called "View" as other programs seem to do;
- This needs Checkboxes for every item in that menu. Visual Studio Code for example:
<img width=50% src=https://github.com/godotengine/godot/assets/66727710/9ed6cecb-ca95-4175-8f43-ff2b04c665b9>

KoBeWi commented:
> The menu could be called "View" as other programs seem to do;

In GIMP the menu is called "Windows".

I think this doesn't really deserve a separate top level menu. It could be inside Editor, e.g. under the layout button.

kitbdev commented:
Rebased since https://github.com/godotengine/godot/pull/88003 is merged.
I moved it into the Editor menu next to the layouts.
Since its no longer top level I changed the name to 'Editor Docks', rather than 'View' or 'Window'.

Incorporated the shortcut, but the shortcut name should probably be changed.

Added checkboxes. This helps see what is closed and what is not.
Originally I wanted the action on clicking it to just open and focus on the dock, but the checkbox implies that clicking it will close the dock, so I'll probably change the functionality to that.
When focusing on the dock, the 2d/3d editor will probably steal the focus anyway.

![image](https://github.com/godotengine/godot/assets/10054226/019fbd36-d9c1-4eaa-bc43-b13aa38d7391)

KoBeWi commented:
Yeah checkbox totally implies that you can toggle it, so it should display differently. You could make open docks show with bold text or using accent color, or show some other icon than checkbox. Maybe eye icon for open docks (to imply focus) and "open" icon for closed docks, to imply opening?

As for the shortcut, you could make the item display the dock's name instead of shortcut name. I think it's fine to reuse it, because it works correctly (i.e. reopens/focuses dock on the side and toggles dock on bottom).

kitbdev commented:
I don't think I can make them bold or colored, PopupMenu items don't support BBCode and they share the same theme. The open eye icon (`GuiVisibilityVisible`) makes it look like clicking will make it hidden, since that is what it does in the Scene Dock. The only open icon I could find was file open (`Load`) but that didn't seem right for opening docks.

Here's what it looks like with `Search` icons for open docks (to 'find' them) and closed eye icons for closed ones.
I also made docks disabled in EditorFeatures be disabled, and set the name for the shortcut.
I tried adding tooltips, but #73926 is causing issues, so I'll remove them until that is fixed.

![image](https://github.com/godotengine/godot/assets/10054226/da3ed08e-19a0-42d8-b333-3a5f3cf24db7)

Let me know if there are better icons or if I should try anything else.
Or I can make it toggle open/closed and use checkboxes.

As for what other engines do: Unity has the menu items always open a new instance of the dock. Unreal has checkboxes to open and close docks, and they are in submenus for a fixed number of multiple instances.

KoBeWi commented:
Hmmm, in GIMP the docks menu simply displays the list of docks with their own icons, without providing information whether they are currently opened. Not sure how it is in other software.
Unique icon per dock is obviously overkill (at least for now), but maybe they could all use the same icon and for displaying status, icon modulate can be used (with inactive docks being gray). As for what icon to use, I think Window is fine.

kitbdev commented:
Here is how it looks with the Window icon with modulate:

![image](https://github.com/godotengine/godot/assets/10054226/a90a8ac3-cc31-4f47-862a-80b082b1765e)

KoBeWi commented:
IMO looks fine.

Close button needs icon:
![image](https://github.com/godotengine/godot/assets/2223172/6246a3e2-9303-48c5-a0ad-436aa68cb709)

KoBeWi commented on doc/classes/EditorPlugin.xml:
```diff
@@ -820,6 +820,9 @@
 			Tab of Project Settings dialog, to the right of other tabs.
 		</constant>
+		<constant name="DOCK_SLOT_NONE" value="-1" enum="DockSlot">
+			Dock slot, none. The dock will start closed and will open as a floating window.
```
This doesn't work that well. When you add a dock as NONE, there is no way to dock it into the editor, as closing the floating window will close the dock.

kitbdev replied:
- This will need something like https://github.com/godotengine/godot-proposals/issues/1984

I can not expose it until then.

KoBeWi replied:
I assumed so. It's better to not expose an effectively broken feature.

kitbdev commented:
Unexposed dock slot none and added close icon:

![image](https://github.com/godotengine/godot/assets/10054226/02cee25b-6cd0-4791-9b62-b5961686307d)

Calinou commented on editor/editor_dock_manager.cpp:
```diff
@@ -201,6 +240,12 @@ void EditorDockManager::_open_dock_in_window(Control *p_dock, bool p_show_window
 		_update_layout();
+		if (p_reset_size) {
+			// Use a default size of one third the current window size.
+			Size2i pupup_size = EditorNode::get_singleton()->get_window()->get_size() / 3.0;
+			p_dock->get_window()->set_size(pupup_size);
```
Typo?

```suggestion
			Size2i popup_size = EditorNode::get_singleton()->get_window()->get_size() / 3.0;
			p_dock->get_window()->set_size(popup_size);
```

Calinou reviewed (commented):
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/29b3d9e9e538f0aa8effc8ad8bf19a2915292a89), it works as expected.

Code looks good to me as well (outside of the suggestions).

Calinou commented on editor/editor_dock_manager.cpp:
```diff
@@ -149,17 +149,56 @@ void EditorDockManager::_update_layout() {
+	docks_menu->reset_size();
+
+	const Ref<Texture2D> icon = docks_menu->get_editor_theme_icon(SNAME("Window"));
+	const Color open_icon_color_mod = Color(1, 1, 1);
+	const Color closed_icon_color_mod = Color(0.5, 0.5, 0.5);
```
This will look strange on a light theme (here, the first dock is closed and other docks are open):

![image](https://github.com/godotengine/godot/assets/180032/4bd68a3d-9e2c-46b8-a939-58fc2f3fec8c)

To avoid this, I suggest multiplying opacity by 0.5 instead of multiplying the color.

kitbdev commented:
Updated to use opacity.

![Screenshot 2024-03-29 215143](https://github.com/godotengine/godot/assets/10054226/c0f5aea1-1380-4ac3-80cf-2e8ca49da968)

Also, I realized that changing the theme wasn't updating it so I connected it to the theme_changed signal.

KoBeWi commented:
You should use `NOTIFICATION_THEME_CHANGED` instead of signal.

KoBeWi commented:
> EditorDockManager doesn't have get that notification

Then signal is fine.

akien-mga commented:
Thanks!

ModuleCode commented:
I think the menu bar should be made to stay after clicking, otherwise it will become very troublesome

kitbdev commented:
Currently, clicking on the menu for an open dock only focuses on the dock, so the menu should close in that case. 
But for opening closed docks, we can make it stay open.

- made a PR for it #91484

RedMser commented on editor/editor_dock_manager.cpp:
```diff
@@ -149,17 +149,53 @@ void EditorDockManager::_update_layout() {
+		if (!dock.value.open) {
+			docks_menu->set_item_icon_modulate(id, closed_icon_color_mod);
+		}
+		docks_menu->set_item_disabled(id, !dock.value.enabled);
```
Features that are disabled via Feature Profiles usually get **hidden** in the UI, not disabled. Seems like a simple change to skip adding the menu item in this case, instead of disabling.

kitbdev replied:
Sorry I forgot about this for a while. 
- Opened PR https://github.com/godotengine/godot/pull/99550
